### PR TITLE
ADD expires, count and lastNotification in NGSIv2 subscription rendering

### PR DESCRIPTION
**Pull request by fgalan**
Part of #1126 implementation.

### CODE REVIEW

File: doc/manuals/admin/database_model.md [hunk 1]
```diff
@@ -86,55 +86,55 @@ Fields:
 Example document:
 
 ```
- {
```
Comment (fgalan) [inline doc/manuals/admin/database_model.md]: My IDE has ruin the whole formatting (only one line was touched). I will fix it.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 2]
```diff
@@ -855,7 +855,10 @@ A subscription is represented by a JSON object with the following fields:
 + `id`: Subscription unique identifier. Automatically created at creation time.
 + `subject`: It is an object that describes the subject of the subscription.
 + `notification`: It is an object that describes the notification received by the subscriber.
-+ `duration`: Duration of the subscription in ISO8601 format. Infinite if not specified.
++ `expires`: Subscription expiration date in ISO8601 format. Permanent subscriptions must omit
+   this field.
++ `status` (not editable, only in GET operations): Either `active` (for permanent subscriptions or if
+   the subscription hasn't expired yet) or `expired` (for expired subscriptions).
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: nit: has not

Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: maybe in the future we could have a 'suspended' status as well.

File: src/lib/common/JsonHelper.cpp [hunk 3]
```diff
@@ -149,6 +154,46 @@ void JsonHelper::addRaw(const std::string& key, const std::string& value)
...
+* JsonHelper::addDate -
+*/
+void JsonHelper::addDate(const std::string& key, long long timestamp)
+{
+  if (!empty)
+  {
+    ss << ',';
+  }
+
+  // 80 bytes are large enough to store any ISO8601 string safely
```
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: "1970-04-26T17:46:40.00Z" is 24 char long. Event assuming some more space for the timezone in the future, I think that 80 bytes will suffice al cases.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 4]
```diff
@@ -855,7 +855,10 @@ A subscription is represented by a JSON object with the following fields:
 + `id`: Subscription unique identifier. Automatically created at creation time.
 + `subject`: It is an object that describes the subject of the subscription.
 + `notification`: It is an object that describes the notification received by the subscriber.
-+ `duration`: Duration of the subscription in ISO8601 format. Infinite if not specified.
++ `expires`: Subscription expiration date in ISO8601 format. Permanent subscriptions must omit
+   this field.
++ `status` (not editable, only in GET operations): Either `active` (for permanent subscriptions or if
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: for the moment ok for only in GET operations but I can imagine creating subscriptions which are not initially active ...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 5]
```diff
@@ -855,7 +855,10 @@ A subscription is represented by a JSON object with the following fields:
 + `id`: Subscription unique identifier. Automatically created at creation time.
 + `subject`: It is an object that describes the subject of the subscription.
 + `notification`: It is an object that describes the notification received by the subscriber.
-+ `duration`: Duration of the subscription in ISO8601 format. Infinite if not specified.
++ `expires`: Subscription expiration date in ISO8601 format. Permanent subscriptions must omit
+   this field.
```
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: ISO8601 is a very wide format.
We will only accept exact date, including hour, minute and seconds, and ONLY UTC is to be used. Right?
This is very important information and must be documented. 

Emphasize on that it is a **date**, not a duration.

For example, "P1M" is ISO8601 but will not be accepted.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 6]
```diff
@@ -873,8 +876,10 @@ A `notification` object contains the following subfields:
   in the notification
 + `callback` : URL pointing to the service which will be invoked when a notification is generated. A NGSIv2
   compliant server must support `http` URL schema, other schemas (e.g. schemas for web sockets) could also be supported.
-+ `throttling`: Minimal period of time (in ISO8601 format) which must elapse between two consecutive notifications. It
++ `throttling`: Minimal period of time in seconds which must elapse between two consecutive notifications. It
   is optional.
++ `count` (not editable, only in GET operations): Number of notifications sent due to this subscription.
++ `lastNotification`: Last notification date in ISO6601 format.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: lastNotification seems to be a bit redundant with the field 'notification' . I think we can find another name for it ... but I do not have a good one right now

Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: 8601?  :-)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 7]
```diff
@@ -873,8 +876,10 @@ A `notification` object contains the following subfields:
   in the notification
 + `callback` : URL pointing to the service which will be invoked when a notification is generated. A NGSIv2
   compliant server must support `http` URL schema, other schemas (e.g. schemas for web sockets) could also be supported.
-+ `throttling`: Minimal period of time (in ISO8601 format) which must elapse between two consecutive notifications. It
++ `throttling`: Minimal period of time in seconds which must elapse between two consecutive notifications. It
   is optional.
++ `count` (not editable, only in GET operations): Number of notifications sent due to this subscription.
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Could it be ```timesSent``` ?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 8]
```diff
@@ -896,6 +901,13 @@ Response:
 * Successful operation uses 200 OK
 * Errors use a non-200 and (optionally) an error payload. See subsection on Error Responses for more details.
 
++ Parameters
+    + limit: 10 (optional, number) - Limit the number of types to be retrieved
+    + offset: 20 (optional, number) - Skip a number of records
+    + options (optional, string) - Options dictionary
+      + Members
+          + count - when used, the total types number is returned as a HTTP in the response named `X-Total-Count`
```
Comment (jmcanterafonseca) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: cut & paste problem. It should say total subscription number

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Also, would not "expirationDate" be a **much** better name than "expires"?

File: src/lib/apiTypesV2/Subscription.cpp [hunk 9]
```diff
@@ -63,10 +64,18 @@ std::string Notification::toJson()
...
+    jh.addLongLong("throttling", this->throttling);
+  }
+  if (this->count > 0)
+  {
+    jh.addLongLong("count", this->count);
+  }
+  if (this->lastNotification > 0)
+  {
+    jh.addDate("lastNotification", this->lastNotification);
   }
```
Comment (kzangeli) [inline src/lib/apiTypesV2/Subscription.cpp]: An empty line between every if () {}  makes the function easier to read.
Not yet in style-guide, but it will come (as SHOULD, not MUST)

File: src/lib/common/JsonHelper.cpp [hunk 10]
```diff
@@ -149,6 +154,46 @@ void JsonHelper::addRaw(const std::string& key, const std::string& value)
...
+  // We use gmtime() to get UTC strings, otherwise we would use localtime()
+  // Date pattern: 1970-04-26T17:46:40.00Z
+  char buffer[80];
+  time_t rawtime = (time_t) timestamp;
```
Comment (kzangeli) [inline src/lib/common/JsonHelper.cpp]: style-guide: align vars horizontally

File: src/lib/common/JsonHelper.cpp [hunk 11]
```diff
@@ -149,6 +154,46 @@ void JsonHelper::addRaw(const std::string& key, const std::string& value)
...
+  strftime (buffer, sizeof(buffer),"%Y-%m-%dT%H:%M:%S.00Z", gmtime(&rawtime));
```
Comment (kzangeli) [inline src/lib/common/JsonHelper.cpp]: style-guide: no space between function name and '('

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: lastNotificationTime

File: src/lib/common/JsonHelper.cpp [hunk 12]
```diff
@@ -149,6 +154,46 @@ void JsonHelper::addRaw(const std::string& key, const std::string& value)
 }
 
 
+/* ****************************************************************************
+*
+* JsonHelper::addDouble -
```
Comment (crbrox) [inline src/lib/common/JsonHelper.cpp]: addLongLong ?

Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: Let's use addNumber at the end.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: As far as I understand, the exact encoding is part of the ISO8601 string, so it can be univocally identified (or generated) by the client. I mean, if you use:

2016-04-05T14:00:00.00Z

you are meaning a date in UTC format while if you use

1994-11-05T08:15:30-05:00

you are referring to a time in GMT-5 timezone.

So, no ambiguety is possible within the ISO8601 standard.

Moreover, currently Orion generates dates in UTC but 1) this may change in the future, 2) we haven't defined yet what to do with dates provided by clients not using UTC (and maybe it is easy to process them if the library we use from parsing convert them to the right timestamp time_t for free). Moreover, both can be considered subtle detailt related with our particular NGSIv2 implementation.

Bottom line: I think is ok have the current wording in the spec. However, as I understand the whole specification will be reviewed once NGSIv2 get fully implemented, we have an opportunity in the future to change it if it doesn't match the final implementation in Orion.

Regarding emphasizing "date", I also think that it is pretty clear with the current working. Adding more stuff could make understanding harder (*excusatio non petita accusatio manifesta* ;).

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: It would depend on how https://github.com/telefonicaid/fiware-orion/issues/1328 gets implemented at the end. Could you add a comment about this in that issue, pls?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Regarding "expirationDate", I like "expires" as it is shorter, but let's discuss offline about it...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: My idea was adjusting the .apib to add the new status when we address https://github.com/telefonicaid/fiware-orion/issues/1328. Some uncertainty exist at the present moment.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 7 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Yes, probably it is better... "count" is the same work we are using for one of the options, so it could misselead readers.

File: src/lib/common/JsonHelper.cpp (hunk 12 above)
Comment (kzangeli) [inline src/lib/common/JsonHelper.cpp]: Sounds good.
Might have to add a

    addNumber(const std::string& key, double value)

in the future?

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: So. I might have misunderstood this.
If this part refers to a **response** from the broker, then I agree.
I got the impression (don't know from where) that it was a **request** being described.
Your comments make me believe it's a response ...

File: src/lib/common/JsonHelper.cpp (hunk 12 above)
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: That's the idea... to add new addNumber() functions for other kind of number types if we need in the future.

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Actually, it refers to both (JSON definition is common... maybe looking the whole document could help to see how it fits: http://telefonicaid.github.io/fiware-orion/api/v2/).

Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: In that case, I would insist on clearly explaining that "expires" needs to be an exact date, that "P1M" is ISO8601, but the broker will not accept it.

Also, as discussed somewhere, the date must be complete with hour, minute and second (that *might* be part of ISO8601, but still, it is never a bad idea to explain it).

About "expires" vs "expirationDate" - "expires" gives me the feeling of a boolean. "expirationDate" on the other hand is very clear. It also removes the doubt about using "P1M", which is not a date, but a period of time.

File: src/lib/apiTypesV2/Subscription.cpp (hunk 9 above)
Comment (fgalan) [inline src/lib/apiTypesV2/Subscription.cpp]: I think in this case (wich one line statement, very regular among them, all using add*() method), having in this way is better. Thus, I'm applying the SHOULD ;)

NTC

Comment (kzangeli) [inline src/lib/apiTypesV2/Subscription.cpp]: Actually, this is a pretty good candidate for the one-liner ifs (if you want to save lines and improve readibility). Look:

    if (this->throttling > 0)           { jh.addLongLong("throttling", this->throttling);         }
    if (this->count > 0)                { jh.addLongLong("count", this->count);                   }
    if (this->lastNotification > 0)     { jh.addDate("lastNotification", this->lastNotification); }

Just a suggestion.

File: src/lib/common/JsonHelper.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: NTC

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 4 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: NTC

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 5 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: When the referred standard describes the behavior I think the best approach is not including the details in the document pointing to the standard for two main reasons:

* Avoid potential contractions (noboby will provide a better description than the standard itself)
* Avoid making the document unnecesarry long or complex

In this particular case, I think is pretty clear: use ISO8601 dates. On the one hand, within the ISO8601 specification I think is pretty clear what a date is and what a interval (duration) is. On the other hand, think we are very "biassed" to the "P1M" notation because we have suffered a long time :) but most of poeple doesn't think in duration when the think ISO8601. Thus, I don't see the need to stress this.

However, it is not so important in this PR to have a perfect wording for this, as I said in a previous comment. We don't know yet all the details on how the ISO8601 stuff will be implemented at the end.

Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: OK, I give up.

NTC

File: doc/manuals/admin/database_model.md (hunk 1 above)
Comment (fgalan) [inline doc/manuals/admin/database_model.md]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed (the nit) in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 6 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed (the ISO number in) 18aa61203ef827009f1f4d4b64019cb6025110b8

Regarding field name, lastNotification was decided at the end (see skype chat discussion).

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 7 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 8 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: src/lib/common/JsonHelper.cpp (hunk 10 above)
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: src/lib/common/JsonHelper.cpp (hunk 11 above)
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

File: src/lib/common/JsonHelper.cpp (hunk 12 above)
Comment (fgalan) [inline src/lib/common/JsonHelper.cpp]: Fixed in 18aa61203ef827009f1f4d4b64019cb6025110b8

**Comment by fgalan** [comment]
I have needed to put the test related with "get a particular subscription" in .DISABLED in this PR, but it will be recovered very soon (when the "get a particular subscription" mongoBackend part starts to be implemented).

**Comment by crbrox** [comment]
LGTM

**Comment by kzangeli** [comment]
LGTM
